dali/pipeline/operator/sequence_operator.h

```diff
@@ -0,0 +1,451 @@
+    return expanded_.template Input<InputBackend>(intput_idx);
+  }
+
+  template <typename OutputBackend>
+  const auto &ExpandedOutput(int output_idx) {
+    return expanded_.template Output<OutputBackend>(output_idx);
+  }
+
+  void Clear() {
+    expanded_.Clear();
+    input_expand_desc_.clear();
+  }
+
+  TensorVector<CPUBackend> ExpandArgumentLike(const TensorVector<CPUBackend> &arg_tensor,
+                                              const std::string &arg_name,
+                                              const ExpandDesc &expand_desc) {
+    DALI_ENFORCE(static_cast<ptrdiff_t>(arg_tensor.num_samples()) == expand_desc.NumSamples(),
+                 make_string("Number of samples passed for argument ", arg_name, " (got ",
+                             arg_tensor.num_samples(),
+                             ") does not match the number of samples in the input (got ",
+                             expand_desc.NumSamples(), ")."));
+    const auto &arg_layout = arg_tensor.GetLayout();
+    int arg_sample_dim = arg_tensor.sample_dim();
+    DALI_ENFORCE(
+        arg_layout.size() == 0 || arg_sample_dim == arg_layout.size(),
+        make_string("Layout of argument input ", arg_name, " has size ", arg_layout.size(),
+                    " which does not match the dimensionality of the argument (got argument "
+                    "with dimensionality ",
+                    arg_sample_dim, ")."));
+    if (!IsPerFrame(arg_tensor)) {
+      return BroadcastPerFrameLike(arg_tensor, expand_desc);
+    }
+    return ExpandPerFrameLike(arg_tensor, arg_name, expand_desc);
+  }
+
+  TensorVector<CPUBackend> ExpandPerFrameLike(const TensorVector<CPUBackend> &arg_tensor,
+                                              const std::string &arg_name,
+                                              const ExpandDesc &expand_desc) {
+    DALI_ENFORCE(
+        expand_desc.HasFrames(),
+        make_string(
+            "Tensor input for argument ", arg_name, " is specified per frame (got ",
+            arg_tensor.GetLayout(),
+            " layout), but samples in the input batch do not contain frames (expected input "
+            "layout that starts with 'F', ",
+            ShouldExpandChannels() ? " or 'CF'" : "", ")."));
+    const auto &shape = arg_tensor.shape();
+    int arg_sample_dim = shape.sample_dim();
+    TensorVector<CPUBackend> flat_tensor(expand_desc.NumElements());
+    int num_samples = expand_desc.NumSamples();
+    auto type_info = arg_tensor.type_info();
+    auto type = type_info.id();
+    auto is_pinned = arg_tensor.is_pinned();
+    auto order = arg_tensor.order();
+    int elem_idx = 0;
+    for (int sample_idx = 0; sample_idx < num_samples; sample_idx++) {
+      const auto &sample_shape = shape[sample_idx];
+      int num_input_frames = expand_desc.NumFrames(sample_idx);
+      int num_arg_frames = sample_shape[0];
+      DALI_ENFORCE(
+          num_arg_frames == 1 || num_input_frames == num_arg_frames,
```

Comment:
Not a suggestion but rather a question: should we broadcast length-1 per-frame arguments? On one hand, it seems natural to require an exact match of frame count if the argument is marked as per-frame - on the other, expanding single-frame to multi-frame follows NumPy-like logic.
What do you think?

Reply:
I thought that with general effort to support broadcasting, this place should not be an exception from the rule. And it does not require much effort on DALI side.